File: java/course-repo/solution/03-dedup-via-id-reuse/src/main/java/webhook/SendDouble.java
package webhook;

import io.temporal.api.enums.v1.ActivityIdConflictPolicy;
import io.temporal.client.ActivityClient;
import io.temporal.client.ActivityClientOptions;
import io.temporal.client.ActivityHandle;
import io.temporal.client.StartActivityOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import java.time.Duration;
import java.util.Map;

public class SendDouble {
    static ActivityHandle<Integer> start(ActivityClient client, String eventId, String label) {
        System.out.println("[" + label + "] start activityId=deliver-" + eventId);
        try {
            StartActivityOptions options = StartActivityOptions.newBuilder()
                    .setId("deliver-" + eventId)
                    .setTaskQueue(Shared.TASK_QUEUE)
                    .setStartToCloseTimeout(Duration.ofSeconds(30))
                    // USE_EXISTING: the second submit with the same ID returns a handle to the
                    // existing Activity instead of erroring. Server-side dedup before any Worker runs it.
                    .setIdConflictPolicy(ActivityIdConflictPolicy.ACTIVITY_ID_CONFLICT_POLICY_USE_EXISTING)
                    .build();
            ActivityHandle<Integer> handle = client.start(
                    WebhookActivities.class, WebhookActivities::deliverWebhook, options,
                    new WebhookDelivery(Shared.WEBHOOK_RECEIVER_URL,
                            Map.of("eventId", eventId, "type", "dup_test"), eventId));
            System.out.println("[" + label + "] handle ok (activityId=deliver-" + eventId + ")");
            return handle;
        } catch (Exception e) {
            System.out.println("[" + label + "] FAILED: " + e.getMessage());
            return null;
        }
    }

    public static void main(String[] args) {
        String eventId = args.length > 0 ? args[0] : "evt_dup";

        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        ActivityClient client = ActivityClient.newInstance(
                service, ActivityClientOptions.newBuilder().setNamespace("default").build());

        ActivityHandle<Integer> h1 = start(client, eventId, "call-1");
        ActivityHandle<Integer> h2 = start(client, eventId, "call-2");
        if (h1 != null) {
            h1.getResult();
            System.out.println("[call-1] activity completed");
        }
        if (h2 != null) {
            h2.getResult();
            System.out.println("[call-2] activity completed");
        }
    }
}

File: typescript/course-repo/solution/03-dedup-via-id-reuse/src/sendDouble.ts
/**
 * Call start() TWICE with the same activityId, back-to-back.
 *
 * The second call uses ActivityIdConflictPolicy.USE_EXISTING so it
 * quietly returns the existing handle instead of erroring.
 */
import { ActivityIdConflictPolicy, Client, Connection } from '@temporalio/client';
import { TASK_QUEUE, WEBHOOK_RECEIVER_URL } from './shared';

async function start(client: Client, eventId: string, label: string) {
  console.log(`[${label}] start activityId=deliver-${eventId}`);
  try {
    const handle = await client.activity.start("deliverWebhook", {
      args: [{ url: WEBHOOK_RECEIVER_URL, payload: { eventId, type: 'dup_test' }, eventId }],
      id: `deliver-${eventId}`,
      taskQueue: TASK_QUEUE,
      startToCloseTimeout: '30 seconds',
      // USE_EXISTING: second submission with the same activityId returns the
      // existing handle instead of erroring. Server-side dedup before any Worker sees the job.
      idConflictPolicy: ActivityIdConflictPolicy.USE_EXISTING,
    });
    console.log(`[${label}] handle ok (activityId=${handle.activityId})`);
    return handle;
  } catch (err) {
    console.log(`[${label}] FAILED: ${err}`);
    return null;
  }
}

async function main(eventId: string) {
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const h1 = await start(client, eventId, 'call-1');
  const h2 = await start(client, eventId, 'call-2');

  if (h1) { await h1.result(); console.log('[call-1] activity completed'); }
  if (h2) { await h2.result(); console.log('[call-2] activity completed'); }

  await connection.close();
}

const eventId = process.argv[2] ?? 'evt_dup';
main(eventId).catch((err) => {
  console.error(err);
  process.exit(1);
});
